### ISSUE TITLE
Deepspeed support CosineAnnealingLR scheduler

### ISSUE THREAD

**Issue by hahchenchen** [microsoft/DeepSpeed#3492]
Now Deepspeed do not support CosineAnnealingLR scheduler.

So i want to support it by meself. question is how to develop a custom scheduler, Are there any tutorials available?

**Comment by DavdGao** [comment]
The same question for me

**Comment by MarshtompCS** [comment]
same issue using stage-2

**Comment by tjruwase** [comment]
@hahchenchen and @DavdGao, unfortunately we don't have a tutorial for this. However, there are two options available for this issue:
1. You can directly pass the torch [implementation](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingLR.html) into `deepspeed.initialize()` as documented [here](https://deepspeed.readthedocs.io/en/latest/initialize.html#training-initialization).
2. You can implement directly in deepspeed by following existing custom [implementations](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/lr_schedules.py), such as [WarmupLR](https://github.com/microsoft/DeepSpeed/blob/ff7d5275f2aa916cb5f320e0d817154e96f9cdb6/deepspeed/runtime/lr_schedules.py#L626) or [OneCycle](https://github.com/microsoft/DeepSpeed/blob/ff7d5275f2aa916cb5f320e0d817154e96f9cdb6/deepspeed/runtime/lr_schedules.py#L361).

**Comment by HsuWanTing** [comment]
Hi @tjruwase , 

> You can directly pass the torch [implementation](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingLR.html) into deepspeed.initialize() as documented [here](https://deepspeed.readthedocs.io/en/latest/initialize.html#training-initialization).

For this option, do you mean like the following?

```
model, optimizer, _, lr_scheduler = deepspeed.initialize(
        model=model,
        args=args,
        lr_scheduler=torch.optim.lr_scheduler.CosineAnnealingLR,
        config_params=ds_config
    )
```

If yes, there is a `T_max` argument in `CosineAnnealingLR` that is required. How to pass this argument?

**Comment by tjruwase** [comment]
@HsuWanTing, you can also pass lr scheduler as a Callable, which should work for your case. Please see the following example
https://github.com/microsoft/DeepSpeed/blob/3dd7ccff8103be60c31d963dd2278d43abb68fd1/tests/unit/runtime/test_ds_initialize.py#L254
